File: singly-linked-list/sll.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    
    def print_list(self):
        node = self.head
        nodes = []
        if node is None: return print("This list has no nodes!")
        while node != None:
            nodes.append(node.value)
            node = node.next
        print(nodes)
    
    def push(self, value):
        newNode = Node(value)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            self.tail = newNode
        self.length += 1
        return self
    
    def pop(self):
        popped = None
        if self.length == 0:
            return popped
        elif self.length == 1:
            popped = self.tail
            self.head = None
            self.tail = None
        else:
            newTail = self.head
            popped = newTail.next
            while popped.next:
                newTail = newTail.next
                popped = newTail.next
            self.tail = newTail
            self.tail.next = None
        self.length -= 1
        return popped
    
    def unshift(self, value):
        newNode = Node(value)
        if self.head == None:
            self.head = newNode
            self.tail = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.length += 1
        return self
    
    def shift(self):
        shifted = None
        if self.head == None:
            return shifted
        elif self.length == 1:
            shifted = self.head
            self.head = None
            self.tail = None
        else:
            shifted = self.head
            self.head = shifted.next
            shifted.next = None
        self.length -= 1
        return shifted
    
    def get(self, position):
        if position < 0 or position >= self.length:
            return None
        i = 0
        target = self.head
        while i != position:
            target = target.next
            i += 1
        return target

    def set(self, value, position):
        target = self.get(position)
        if target != None:
            target.value = value
            return True
        return False

    def insert(self, value, position):
        if position < 0 | position > self.length:
            return None
        elif position == self.length:
            return self.push(value)
        elif position == 0:
            return self.unshift(value)
        newNode = Node(value)
        prev = self.get(position - 1)
        next = prev.next
        prev.next = newNode
        newNode.next = next
        self.length += 1
        return self
    
    def remove(self, position):
        if position < 0 | position >= self.length:
            return None
        elif position == self.length - 1:
            return self.pop()
        elif position == 0:
            return self.shift()
        prev = self.get(position - 1)
        target = prev.next
        next = target.next
        prev.next = next
        target.next = None
        self.length -= 1
        return target
    
    def reverse(self):
        node = self.head
        self.head = self.tail
        self.tail = node
        prev = None
        next = None
        while node:
            next = node.next
            node.next = prev
            prev = node
            node = next
        return self


my_list = SinglyLinkedList()
my_list.push("A")
my_list.push("B")
my_list.push("C")
my_list.push("D")
my_list.push("E")
# # my_list.pop()
# my_list.unshift("mondir")
# my_list.shift()
# my_list.set("mondir", 4)

# print(my_list.set("ahmed",3))
# my_list.print_list()
# my_list.insert("yasser", 0)
my_list.reverse()
my_list.print_list()
# print(my_list.get(4))
# print(my_list.length, my_list.head.value, my_list.tail.value)

File: singly-linked-list/sll.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class SLL {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        let newNode = new Node(value);
        if (this.head === null || this.length === 0) {
            this.head = newNode;
            this.tail = this.head;
        }
        else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    pop() {
        if (!this.length) return undefined;
        else if (this.head === this.tail) {
            let popped = this.head;
            this.head = null;
            this.tail = null;
            this.length--;
            return popped;
        }

        let newTail = this.head;
        let oldTail = newTail.next;
        while (oldTail.next) {
            newTail = newTail.next;
            oldTail = newTail.next;
        }
        this.tail = newTail;
        this.tail.next = null;
        this.length--;
        return oldTail;
    }

    unshift(value) {
        let newNode = new Node(value);
        if (this.head == null) {
            this.head = newNode;
            this.tail = this.head;
        }
        else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        return this;
    }

    shift() {
        let shifted, newHead;
        if (this.length === 0) {
            shifted = undefined;
        }
        if (this.head === this.tail) {
            shifted = this.head;
            this.head = null;
            this.tail = this.head;
            this.length--;
        }
        else {
            shifted = this.head;
            newHead = shifted.next;
            shifted.next = null;
            this.head = newHead;
            this.length--;
        }
        return shifted;
    }

    get(position) {
        if (position < 0 || this.length <= position) return undefined;
        let target = this.head;
        let counter = 0;
        while (counter != position) {
            target = target.next;
            counter++;
        }
        return target;
    }

    set(value, position) {
        let target = this.get(position);
        if (target != undefined) {
            target.value = value;
            return true;
        }
        return false;
    }

    insert(value, position) {
        if (position < 0 || position > this.length) return false;
        if (position === 0) {
            this.unshift(value);
            return true;
        }
        if (position === this.length) {
            this.push(value);
            return true;
        }
        // create node to be inserted
        let newNode = new Node(value);
        // find the node that's before it
        let nodeBefore = this.get(position - 1);
        // find the node that's after it
        let nodeAfter = nodeBefore.next;
        // link the node to be intserted to it's next node
        newNode.next = nodeAfter;
        // link the node before it to the new node
        nodeBefore.next = newNode;

        this.length++;
        return true;
    }

    remove(position) {
        if (position < 0 || position >= this.length) return undefined;
        if (position === 0) return this.shift();
        if (position === this.length - 1) return this.pop();
        // find the node before target
        let nodeBefore = this.get(position - 1);
        // find the target
        let target = nodeBefore.next;
        // find the node after the target
        let nodeAfter = target.next;
        // link the before and after together
        nodeBefore.next = nodeAfter;

        this.length--;
        target.next = null;
        return target;
    }

    reverse() {
        // change the head and tail
        let oldHead = this.head;
        this.head = this.tail;
        this.tail = oldHead;
        // set a pointer (before, after)
        let previous = null;
        let next = null;
        for (let i = 0; i < this.length; i++) {
            next = oldHead.next;
            oldHead.next = previous;
            previous = oldHead;
            oldHead = next;
        }
    }
}

module.exports = SLL;
